fix(core): Add timeout and recovery for database connection health checks

despairblue opened the pull request:
## Summary

This PR fixes an issue where n8n would not properly detect and recover from database connection timeouts. When the database becomes unreachable (e.g., network partition), n8n's health checks would hang indefinitely instead of reporting the issue.

**Why it hangs forever:**
When the database connection is interrupted, the pg connection pool returns stale connections that don't know they're broken (TCP doesn't detect network failures immediately). When TypeORM tries to query on these connections, the query hangs at the TCP layer waiting for a response that will never come. The promise never resolves or rejects - it just waits indefinitely.

**Changes:**
1. Added 5-second timeout to database ping queries using Promise.race to prevent indefinite hangs
2. Added automatic recovery detection and logging when database connection is restored
3. Updated health check to properly reflect database connection state in readiness probe
4. Added chaos test to verify timeout detection and automatic recovery

**Testing:**
The chaos test simulates a database network partition using iptables and verifies:
- Database connection timeout is detected within 5 seconds
- Liveness probe remains healthy (server is running)
- Readiness probe correctly reports unhealthy (database unavailable)
- Automatic recovery when database becomes reachable again

## Related Linear tickets, Github issues, and Community forum posts

https://linear.app/n8n/issue/CAT-1018

## Review / Merge checklist

- [x] PR title and summary are descriptive
- [ ] Docs updated or follow-up ticket created
- [x] Tests included (chaos test added)
- [ ] PR Labeled with `release/backport` (if needed)

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds a 5s timeout to DB ping with recovery/timeout logging, streamlines health checks, and introduces chaos tests and helpers to validate behavior.
> 
> - **DB connection (`packages/@n8n/db`)**:
>   - Add `Logger` to `DbConnection` and log on timeout (`warn`) and recovery (`info`).
>   - Implement 5s timeout for `ping()` via `Promise.race` with `timers/promises`, abortable via `AbortController`.
>   - Mark timeouts as `OperationalError`; keep scheduling next ping in `finally`.
>   - Update tests to inject `Logger`, assert scheduled pings, and adapt mocks.
> - **Server (`packages/cli/src/abstract-server.ts`)**:
>   - Make `setupHealthCheck` synchronous and call it without `await`.
>   - Make `onShutdown` synchronous.
> - **Testing**:
>   - Add chaos test `tests/chaos/postgres.spec.ts` to simulate DB network partition and verify timeout, readiness degradation, and recovery logs.
>   - Enable `NET_ADMIN` capability for Postgres test container to manipulate iptables.
>   - Extend container helpers: optional `throwOnTimeout` in `waitForLog`, return `null` on timeout.
>   - Add `api.isHealthy(probe)` helper for liveness/readiness checks.
>   - Minor test robustness updates in multi-main leader election.
>   - Add `@n8n/constants` dev dependency for Playwright tests.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit ef93307d4ea76e9b1b71959a8b0d762c1d9fe056. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

shortstacked committed:
test: Add test for Postgres DB connection issues

despairblue committed:
update test to test for the correct behavior

despairblue committed:
make functions without await sync

despairblue committed:
fix readiness check

despairblue committed:
fix type errors and lint issues

despairblue commented on packages/testing/playwright/tests/chaos/postgres.spec.ts:
```diff
@@ -0,0 +1,65 @@
+		// Block all incoming TCP traffic to PostgreSQL port 5432
+		// This simulates a network partition between n8n and the database
+		const rule = ['INPUT', '-p', 'tcp', '--dport', '5432', '-j', 'DROP'];
+		const blockPostgresTraffic = await postgres.exec(['iptables', '-A', ...rule]);
```
Different tests do not share containers, a new one is spun up for every test.
Looking at the fixture configuration in packages/testing/playwright/fixtures/base.ts:

```
  n8nContainer: [
      async ({ containerConfig }, use) => {
          // ... create container ...
          await use(container);
          await container.stop();  // Always stops after test
      },
      { scope: 'worker', box: true },  // box: true = isolated per test file
  ],
```

despairblue commented on packages/@n8n/db/src/connection/__tests__/db-connection.test.ts:
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable @typescript-eslint/unbound-method */
```
**unrelated (non-blocking):** Should we maybe disable this for test files? 
Or use `eslint-plugin-jest`, like recommended in the [docs of this rule](https://typescript-eslint.io/rules/unbound-method/)?

tomi replied:
Yeah I think that's the way to go. Let's create a ticket about this

despairblue commented on packages/@n8n/db/src/connection/__tests__/db-connection.test.ts:
```diff
@@ -133,6 +138,7 @@ describe('DbConnection', () => {
 		it('should update connection state on successful ping', async () => {
 			// @ts-expect-error readonly property
 			dataSource.isInitialized = true;
+			// eslint-disable-next-line @typescript-eslint/naming-convention
```
**unrelated (non-blocking):** We could also be less strict with naming conventions in test files.

despairblue commented on packages/cli/src/abstract-server.ts:
```diff
@@ -296,7 +296,7 @@ export abstract class AbstractServer {
 	 */
 	@OnShutdown()
-	async onShutdown(): Promise<void> {
+	onShutdown(): void {
```
Unrelated eslint warning.

despairblue commented on packages/testing/playwright/tests/chaos/multimain.spec.ts:
```diff
@@ -8,10 +8,10 @@ test('Leader election @mode:multi-main @chaostest', async ({ chaos }) => {
+	const currentLeader = findContainerByLog?.containerName;
 	// Stop leader
-	await chaos.stopContainer(currentLeader);
+	await chaos.stopContainer(currentLeader!);
```
Was already on this branch. @shortstacked should I keep this or remove it?

tomi reviewed (dismissed):
👍 just couple minor comments

despairblue commented on packages/@n8n/db/src/connection/db-connection.ts:
cd7562fa95023e5cb486dd423b5549a74167ab59

If it's an OperationalError it will only log a warning, but if it's anything else we still report it to Sentry.

despairblue committed:
throw and OperationalError when the db connection is interrupted and don't report it to sentry

If another error is thrown we still report it to sentry because it could
be a bug.

despairblue committed:
abort the timeout when the database responds

despairblue committed:
handle potential rejection due to race condition

despairblue committed:
Revert "handle potential rejection due to race condition"

This reverts commit bb86ceff9e8bdd785f8e92ee8417c6bf0bd41c89.

tomi reviewed (approved):
🚀
